drivers/net/ethernet/renesas/rswitch_tc_u32.c

```diff
@@ -48,14 +73,274 @@ static void rswitch_init_u32_redirect_action(struct rswitch_tc_filter *cfg,
+
+	used_two_bytes_hw = get_two_byte_filter(priv);
+	used_four_bytes_hw = get_four_byte_filter(priv);
+	relative_two_bytes_used =
+		((used_two_bytes_hw + pf_entries->two_byte) * 100) / PFL_TWBF_N;
+	relative_four_bytes_used =
+		((used_four_bytes_hw + pf_entries->four_byte) * 100) / PFL_FOBF_N;
+
+	if (used_two_bytes_hw > 0 &&
+	    relative_two_bytes_used < relative_four_bytes_used) {
+		return PF_TWO_BYTE;
+	}
+
+	return PF_FOUR_BYTE;
+}
+
+/* Find the least used filter to match three bytes */
+static enum pf_type get_three_byte_matcher(struct rswitch_private *priv,
+					   struct used_pf_entries *pf_entries)
+{
+	int used_three_bytes_hw, used_four_bytes_hw;
+	int relative_three_bytes_used, relative_four_bytes_used;
+
+	used_three_bytes_hw = get_two_byte_filter(priv);
+	used_four_bytes_hw = get_four_byte_filter(priv);
+	relative_three_bytes_used =
+		((used_three_bytes_hw + pf_entries->three_byte) * 100) / PFL_THBF_N;
+	relative_four_bytes_used =
+		((used_four_bytes_hw + pf_entries->four_byte) * 100) / PFL_FOBF_N;
+
+	if (used_three_bytes_hw > 0 &&
+	    relative_three_bytes_used < relative_four_bytes_used) {
+		return PF_THREE_BYTE;
+	}
+
+	return PF_FOUR_BYTE;
+}
+
+/* Find the least used filter to match one or two bytes */
+static enum pf_type get_one_or_two_byte_matcher(struct rswitch_private *priv,
+						struct used_pf_entries *pf_entries)
+{
+	int used_two_bytes_hw, used_three_bytes_hw, used_four_bytes_hw;
+	int relative_two_bytes_used, relative_three_bytes_used, relative_four_bytes_used;
+
+	used_two_bytes_hw = get_two_byte_filter(priv);
+	used_three_bytes_hw = get_three_byte_filter(priv);
+	used_four_bytes_hw = get_four_byte_filter(priv);
+	relative_two_bytes_used =
+		((used_two_bytes_hw + pf_entries->two_byte) * 100) / PFL_TWBF_N;
+	relative_three_bytes_used =
+		((used_three_bytes_hw + pf_entries->three_byte) * 100) / PFL_THBF_N;
+	relative_four_bytes_used =
+		((used_four_bytes_hw + pf_entries->four_byte) * 100) / PFL_FOBF_N;
+
+	if (used_four_bytes_hw >= 0 && relative_four_bytes_used < relative_two_bytes_used &&
+	    relative_four_bytes_used < relative_three_bytes_used) {
+		return PF_FOUR_BYTE;
+	}
+
+	if (used_three_bytes_hw >= 0 && relative_three_bytes_used < relative_two_bytes_used &&
+	    relative_three_bytes_used < relative_four_bytes_used) {
+		return PF_THREE_BYTE;
+	}
+
+	return PF_TWO_BYTE;
+}
+
+static void add_param_entry(struct rswitch_pf_param *param, int offset, uint8_t *val, int len,
+			    struct used_pf_entries *pf_entries)
+{
+	enum pf_type pf_type;
+	u32 val32 = 0, ext_val32 = 0, mask;
+	struct rswitch_private *priv = param->rdev->priv;
+
+	switch (len) {
+	case EIGHT_B:
+		memcpy(&val32, val, sizeof(val32));
+		memcpy(&ext_val32, val + sizeof(val32), sizeof(ext_val32));
+		val32 = ntohl(val32);
+		ext_val32 = ntohl(ext_val32);
+		rswitch_init_expand_pf_entry(param, PF_FOUR_BYTE, val32, ext_val32, offset);
+		pf_entries->four_byte++;
+		break;
+	case SIX_B:
+		val32 = rswitch_mac_left_half(val);
```

Comment:
Should translation to host byte order be here?

Reply:
I changed this conversion to the same way as it is written in the rest places.